### PR TITLE
Implement RFC 53: Low-level I/O primitives.

### PR DESCRIPTION
**Pull request by wanda-phi**
TODO:

- [x] tests

### CODE REVIEW

**Comment by whitequark** [comment]
I think we need also guide-level documentation for instances and IO values,

File: amaranth/hdl/_ast.py [hunk 1]
```diff
@@ -1733,8 +1734,17 @@ def __repr__(self):
                                            self.width, self.stride)
 
 
+class _CatMeta(ABCMeta):
```
Comment (whitequark) [inline amaranth/hdl/_ast.py:1737]: I'd rather make `Cat` a function, renaming the concatenation class to `Concat` and the IO version to `IOConcat` for disambiguation.

Comment (whitequark) [reply amaranth/hdl/_ast.py:1737]: Actually, do you even need a metaclass here? Wouldn't `__new__` suffice?

File: amaranth/hdl/_ast.py [hunk 2]
```diff
@@ -2732,6 +2742,160 @@ def case_repr(keys, stmts):
...
+    def cast(obj):
+        if isinstance(obj, IOValue):
+            return obj
+        elif isinstance(obj, Value) and len(obj) == 0:
+            return IOCat()
+        else:
+            raise TypeError(f"Object {obj!r} cannot be converted to an IO value")
+
+    def __init__(self, *, src_loc_at=0):
+        super().__init__()
```
Comment (whitequark) [inline amaranth/hdl/_ast.py:2756]: There's no superclass, right?

File: amaranth/hdl/_ast.py [hunk 3]
```diff
@@ -2732,6 +2742,160 @@ def case_repr(keys, stmts):
...
+
+    @property
+    def metadata(self):
+        return self._metadata
+
+    def _ioports(self):
+        return {self}
+
+    def __repr__(self):
+        return f"(ioport {self.name})"
```
Comment (whitequark) [inline amaranth/hdl/_ast.py:2817]: Maybe with a dash, to visually group the `io-` variants?

```suggestion
        return f"(io-port {self.name})"
```

File: amaranth/hdl/_ast.py [hunk 4]
```diff
@@ -2732,6 +2742,160 @@ def case_repr(keys, stmts):
...
+    def __init__(self, width, *, name=None, attrs=None, metadata=None, src_loc_at=0):
+        super().__init__(src_loc_at=src_loc_at)
+
+        if name is not None and not isinstance(name, str):
+            raise TypeError(f"Name must be a string, not {name!r}")
+        self.name = name or tracer.get_var_name(depth=2 + src_loc_at)
+
+        self._width = operator.index(width)
+        self._attrs = dict(() if attrs is None else attrs)
+        self._metadata = (None,) * self._width if metadata is None else tuple(metadata)
```
Comment (whitequark) [inline amaranth/hdl/_ast.py:2796]: I think this needs a check for the size and type of the provided `metadata`.

File: amaranth/hdl/_ast.py [hunk 5]
```diff
@@ -2732,6 +2742,160 @@ def case_repr(keys, stmts):
...
+    def _ioports(self):
+        return {self}
+
+    def __repr__(self):
+        return f"(ioport {self.name})"
+
+
+@final
+class IOCat(IOValue):
+    def __init__(self, *args, src_loc_at=0):
```
Comment (whitequark) [inline amaranth/hdl/_ast.py:2822]: This class isn't directly callable, right? I'd rather accept `parts` in `__init__` and ditch `flatten`. (The same for the `class Cat:` so they match).

I don't love `flatten` due to its weird semantics and I somewhat prefer not adding more calls to it.

File: amaranth/hdl/_ast.py [hunk 6]
```diff
@@ -2732,6 +2742,160 @@ def case_repr(keys, stmts):
...
+        if stop not in range(-n, n+1):
+            raise IndexError(f"Cannot stop slice {stop} bits into {n}-bit value")
+        if stop < 0:
+            stop += n
+        if start > stop:
+            raise IndexError(f"Slice start {start} must be less than slice stop {stop}")
+
+        super().__init__(src_loc_at=src_loc_at)
+        self._value = value
+        self._start = int(operator.index(start))
```
Comment (whitequark) [inline amaranth/hdl/_ast.py:2870]: This seems completely redundant with `isinstance(start, int)` above.

File: amaranth/hdl/_ast.py (hunk 1 above)
Comment (wanda-phi) [reply amaranth/hdl/_ast.py:1737]: Ok, going with `__new__`

**Review by whitequark — COMMENTED** [review]
I feel that the algorithms in `Design` would be clearer if it was done in one (or two, second one computing LCA) DFS pass. However, this doesn't need to happen now.

LGTM, the only thing left is docs which I will take care of imminently.

**Review by whitequark — APPROVED** [review]
LGTM, great work!

**Pull request merged**